feat: add GameObjects to visualise PlayerBase EffectAreas

Nixfriarr opened the pull request:
Implemented visualisations of PlayerBase EffectAreas with terrain-conforming green mesh planes, togglable using F7 keypress.
![202505~2](https://github.com/user-attachments/assets/0a7e0e73-da12-4fa5-91ab-9089147c37d8)
![202505~4](https://github.com/user-attachments/assets/7eaecd1f-9237-4804-a2b8-509701adc65c)
![203DA5~1](https://github.com/user-attachments/assets/a2e8ce61-8d70-44a0-918f-0cd57184b0bd)

Nixfriarr committed:
feat: add GameObjects to visualise PlayerBase EffectAreas

Implemented visualisations of PlayerBase EffectAreas with terrain-conforming green mesh planes, togglable using F7 keypress.

searica commented:
That's an interesting approach. I'll have to review it a bit more to get a better sense of it performance-wise and to understand the rendering. I know you commented elsewhere that it seemed to work well. I have a few test worlds with honestly ridiculous numbers of player base areas though, so I usually use that as the benchmark.

Initial comments for adjustments would be that the key to toggle visibility should be configurable, and the color of the area should also be configurable. 

Another issue is that the player-base area is a sphere, not a cylinder, and as far I can tell this does not account for vertical distances? If it does not then it will display incorrect information to the player. That's the main reason I shifted towards visualizing the area as a sphere rather than a ground overlay.

Nixfriarr commented:
Have a play with it and see how you find it. I've used it to spawnproof our server's main town, which has a lot of playerbase pieces, and that was no more laggy than normal. I believe the performance impact is low because it uses the simplest shader with no transparency and no back-face, and while each mesh has quite a few vertices, it's saved as one gameobject so it doesn't have much of an impact once it's generated. 

I thoroughly explored the sphere idea, but I found it was much harder to spot gaps between spheres than a surface at ground level. My implementation exactly matches the 'marching ants' circle around build pieces like the crafting table, [which is a cylinder not a sphere](https://www.reddit.com/r/valheim/comments/lp8lro/is_the_work_benchs_radius_a_sphere_or_a_cylinder/), and that's what players are used to using for spawnproofing in vanilla. 

Just so you know, neither decals nor projection are possible in Valheim - I've tried a lot. The circle shown when holding the hoe is flat, and just the centre point goes down to the terrain. The aforementioned circle around the crafting table is individual arc-shaped gameobjects which are transformed around the circle and down to the terrain.
![20D1FB~1](https://github.com/user-attachments/assets/d917e90f-1b11-425d-a955-40157cfc6f99)

Fully agree that the colour and the keybind should be configurable. This is just the minimum viable product to allow people to use it right away. SafetyStatus doesnt have user-configurable options at all yet (right?) so that might take me some time to get my head around and implement to provide those options.

searica commented:
Hmm, thanks for the response and the pull request in general!

RE - shape: I agree that visualizing with the sphere gets much more messy. Looking at your implementation, I think it should be doable for me to adjust things to account for vertical distance by dropping vertices that are out of bounds and altering the shape accordingly. 

RE - performance: I would actually change the implementation to take advantage of attaching components to gameobjects and attach the meshes and update methods to each GameObject that has a PlayerBase effect. Makes it easier to handle memory allocation and destruction since the object it's attached to being destroyed will auto-destroy the component that handles rendering the shader. I would also add a static set of mesh vertex positions that simply be translated to the center and then scaled by the radius to reduce repeated computations. 

Then the raycasting to find the terrain height (there is already a vanilla method for this and I would recommend using it) can be used to figure out which portions of the mesh are out-of-range vertically and should be culled. 

RE - decals: Yeah I spent a lot of time working on that once upon a time. While I think it is technically do-able it would require building from the ground up via Unity or implementing custom shaders that are shipped with the mod so it is very much not worth it. 

RE- config options: A lot of my other mods have configuration options, I'd recommend taking a look at those for reference. 

You're welcome to make those changes yourself, or wait until I have the time to integrate your changes (would merge your PR to give you credit) and make the changes myself. Though the culling parts of the mesh based on the 3-d distance using the vertical offset would be a stretch goal as it could end up much more complicated to implement than expected.

Nixfriarr committed:
perf: separate mesh vertex calculation to improve performance

Refactored mesh generation so the base shape is built once and reused. 
Avoids redundant vertex and triangle calculations, significantly improving performance when generating visuals.

Nixfriarr commented:
RE - performance: Excellent suggestion to move the vertices calculation out of the visualisation method. There is now no discernable lag on the initial generation of visuals. I've also switched out the raycasting to use the built-in method. The rest of your suggestion seems like it would be a much bigger refactor - you're welcome to implement it but I haven't.

RE - shape: Your concern about the circle-sphere disparity is valid, but I don't think there would be enough real-world cases to justify correcting for it (guesswork, obviously, but I've included my thought process below). We're matching the vanilla visuals for build/ward area, so changing it might introduce confusion. I suggest to log the disparity in the readme and change functionality if the vanilla visual gets updated. 
-Because circles don't tessellate, players will necessarily overlap PlayerBase pieces when spawnproofing. 
-Steep terrain and terrain covered by a structure prevents spawns, so pieces high up will likely not be spawnable below.
-Even tree-house dwellers will likely have ground-level PlayerBase pieces placed around their build if they want to prevent spawns in a wider area.

Nixfriarr committed:
feat: Add config options for visualisation colour and keybind

Added config options for keybind, colour & transparency. Colour field supports both named colours and hex codes to make user input more intuitive.

Nixfriarr commented:
![image](https://github.com/user-attachments/assets/d3918096-a4a0-45f6-b19f-ccdb56f2ce88)

jarrpa commented:
I just pulled and installed this PR, can confirm that it works for me. The bright green safety circles are tolerable as a first-pass implementation for me, I can't tell you how much of a help it is to have any sort of visual indicator.

Nixfriarr commented:
Glad it worked for you too.

> The bright green safety circles are tolerable as a first-pass

Depending on which commit you downloaded, you might have a config file at \BepInEx\config\Searica.Valheim.SafetyStatus.cfg which will allow you to change the colour and transparency of those circles.

searica commented:
Great, I intend to merge this in a couple of weeks once my current major project is complete and I have the time to implement some of the changes I suggested before releasing it. Thanks for all your work on it!

Nixfriarr commented:
Hang on... @searica can you check whether the PlayerBase EffectArea *is* a sphere? I thought I'd have a crack at marking any vertices in the mesh that were outside the EffectArea, so I built a crafting table high up to test, and the build radius didn't change - it still perfectly matched the visual. (Bear in mind the mesh is 0.25m above the terrain.)
![202506~1](https://github.com/user-attachments/assets/4c42044c-38b0-4cf2-8713-f2bf577b0de6)
![202506~2](https://github.com/user-attachments/assets/75457abc-f5ca-47a3-a72c-44f3b69913d8)
